### PR TITLE
compact: reduce DownsampleRange1 to 6d

### LINKED ISSUE #2413: compact: downsample stopped making progress (compacted blocks migrated from prometheus too small)
**Thanos, Prometheus and Golang version used**:  Thanos image:tag v0.11.0

**Object Storage Provider**:  GCS

**What happened**:  thanos-compact downsample stopped making progress (5m incomplete, no 1h).  The situation has continued for 10 hours.

bucket UI apppears as follows:
Screen Shot 2020-04-11 at 8 39 09 AM

**What you expected to happen**: compactor makes progress towards steady-state of downsampled data

**How to reproduce it (as minimally and precisely as possible)**:

newly deployed compactor job against 5 months of data

compact args:
```
         - "compact"
         - "--log.level=info"
         - "--log.format=logfmt"
         - "--http-address=0.0.0.0:10902"
         - "--objstore.config-file=/etc/config/object-store.yaml"
         - "--data-dir=/var/thanos/compact"
         - "--consistency-delay=30m"
         - "--retention.resolution-raw=0d"
         - "--retention.resolution-5m=0d"
         - "--retention.resolution-1h=0d"
         - "--block-sync-concurrency=20"
         - "--compact.concurrency=1"
         - "--wait"
```

**Full logs to relevant components**:

Compactor iterations have settled to doing no work.  There are no warnings or errors in the logs.  Compaction operations are still happening normally.

Log for latest iteration:
<details>Logs
<p>

```
E 2020-04-10T23:40:59.535995275Z level=info ts=2020-04-10T23:40:59.535727872Z caller=clean.go:50 msg="cleaning of aborted partial uploads done"
 
E 2020-04-10T23:40:59.535910379Z level=info ts=2020-04-10T23:40:59.535671469Z caller=fetcher.go:368 component=block.MetaFetcher msg="successfully fetched block metadata" duration=51.342219ms cached=52 returned=52 partial=0
 
E 2020-04-10T23:40:59.484518859Z level=info ts=2020-04-10T23:40:59.484316539Z caller=clean.go:25 msg="started cleaning of aborted partial uploads"
 
E 2020-04-10T23:40:59.484512229Z level=info ts=2020-04-10T23:40:59.484300199Z caller=retention.go:41 msg="optional retention apply done"
 
E 2020-04-10T23:40:59.484473403Z level=info ts=2020-04-10T23:40:59.484219675Z caller=fetcher.go:368 component=block.MetaFetcher msg="successfully fetched block metadata" duration=61.096018ms cached=52 returned=52 partial=0
 
E 2020-04-10T23:40:59.423193158Z level=info ts=2020-04-10T23:40:59.423100065Z caller=retention.go:20 msg="start optional retention"
 
E 2020-04-10T23:40:59.423178026Z level=info ts=2020-04-10T23:40:59.423057958Z caller=compact.go:324 msg="downsampling iterations done"
 
E 2020-04-10T23:40:59.422756013Z level=info ts=2020-04-10T23:40:59.422495525Z caller=fetcher.go:368 component=block.MetaFetcher msg="successfully fetched block metadata" duration=55.296998ms cached=52 returned=52 partial=0
 
E 2020-04-10T23:40:59.367230069Z level=info ts=2020-04-10T23:40:59.367100699Z caller=compact.go:319 msg="start second pass of downsampling"
 
E 2020-04-10T23:40:59.366817119Z level=info ts=2020-04-10T23:40:59.366579816Z caller=fetcher.go:368 component=block.MetaFetcher msg="successfully fetched block metadata" duration=51.449956ms cached=52 returned=52 partial=0
 
E 2020-04-10T23:40:59.315252887Z level=info ts=2020-04-10T23:40:59.315067975Z caller=compact.go:313 msg="start first pass of downsampling"
 
E 2020-04-10T23:40:59.315221167Z level=info ts=2020-04-10T23:40:59.314961081Z caller=compact.go:910 msg="compaction iterations done"
 
E 2020-04-10T23:40:59.213787324Z level=info ts=2020-04-10T23:40:59.213601901Z caller=compact.go:872 msg="start of compaction"
 
E 2020-04-10T23:40:59.195223946Z level=info ts=2020-04-10T23:40:59.195049536Z caller=compact.go:246 msg="deleting outdated block" block=01E59XWF1XDQFG2AY9TR6TVX6P
 
E 2020-04-10T23:40:59.174106196Z level=info ts=2020-04-10T23:40:59.173935931Z caller=compact.go:246 msg="deleting outdated block" block=01E59Q0N419Y9RAZWG2FKEBEFB
 
E 2020-04-10T23:40:59.156003984Z level=info ts=2020-04-10T23:40:59.155810736Z caller=compact.go:246 msg="deleting outdated block" block=01E59G5090ZVF1AX61E07W50CJ
 
E 2020-04-10T23:40:59.135748608Z level=info ts=2020-04-10T23:40:59.135366491Z caller=compact.go:246 msg="deleting outdated block" block=01E59996ZPQWGGQ03M007CNB4P
 
E 2020-04-10T23:40:59.135736558Z level=info ts=2020-04-10T23:40:59.135351688Z caller=compact.go:864 msg="start of GC"
 
E 2020-04-10T23:40:59.135674222Z level=info ts=2020-04-10T23:40:59.135304544Z caller=fetcher.go:368 component=block.MetaFetcher msg="successfully fetched block metadata" duration=145.544367ms cached=52 returned=52 partial=0
 
E 2020-04-10T23:40:58.989982600Z level=info ts=2020-04-10T23:40:58.989582621Z caller=compact.go:858 msg="start sync of metas"
```

</p>
</details>

**Anything else we need to know**:

I have not yet tried restarting the compact job.  I'll wait up to 1 day for advice in case there is something to inspect on the running instance.

I've ruled out it being a Bucket UI issue by comparing `increase(thanos_compact_downsample_total)` to number of downsampled blocks shown in the UI, and they match as expected.

### PR DESCRIPTION
**Pull request by belm0**
Signed-off-by: John Belmonte <john@neggie.net>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Reduced DownsampleRange1 from 10d to 6d.

The intention is to accommodate blocks migrated from prometheus-server,
which has a different compaction schedule ending at 6.75 days.

Also rename `DownsampleRange{0,1}` to `ResLevel{1,2}MinBlockRange`.

Fixes #2413.

### CODE REVIEW

**Comment by belm0** [comment]
this is really a workaround

I'm open to making this a flag, or contributing larger changes to make the situation better.

I don't think anyone has thought about `--shipper.upload_compacted` regarding the differences in the compaction schedule of prometheus-server vs. thanos-compact.

A proper solution might be for sidecar to rewrite compacted blocks to match compact's expectations before uploading.

slack thread here:  https://cloud-native.slack.com/archives/CK5RSSC10/p1587689525452700

**Comment by bwplotka** [comment]
Definitely, we don't want sidecar to rewrite any block this is because:

* Sidecar have low mem footprint and we want to keep this way
* Yet another place of modifying blocks

The idea overall looks great, let's think how we can do that. I think this might work indeed  :+1: Can we add some tests for this?

**Comment by belm0** [comment]
> Can we add some tests for this?

What kind of tests did you have in mind?

This PR is changing a constant in the source code to accommodate an external system.

Perhaps we can add a runtime assertion that DownsampleRange1 lies between the final two levels in the compactionSet.

**Comment by bwplotka** [comment]
> This PR is changing a constant in the source code to accommodate an external system.

In this case we could make it "just" a bit higher than our previous compaction level so 2d.. maybe `3d`, hm? 

For tests I would suggest adding downsampling cases to our e2e here: https://github.com/thanos-io/thanos/blob/master/test/e2e/compact_test.go#L64

**Comment by belm0** [comment]
> In this case we could make it "just" a bit higher than our previous compaction level so 2d.. maybe 3d, hm?

3d sounds OK

> For tests I would suggest adding downsampling cases to our e2e here: https://github.com/thanos-io/thanos/blob/master/test/e2e/compact_test.go#L64

It's significant that there is no e2e test coverage for downsampling, but trying to expand one of the most complex tests may take me over half a day as a newcomer.  This PR has quite low risk already and having that test wouldn't change much.

The assertion/panic I proposed would add important correctness coverage for the downsampling configuration.  I'd like to try that and leave the existing e2e TODO for now 🙏

**Comment by belm0** [comment]
@bwplotka I'd like to merge this for next release

File: docs/components/compact.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ It is generally not semantically concurrency safe and must be deployed as a sing
 It is also responsible for downsampling of data:
 
 - creating 5m downsampling for blocks larger than **40 hours** (2d, 2w)
-- creating 1h downsampling for blocks larger than **10 days** (2w).
+- creating 1h downsampling for blocks larger than **3 days** (2w).
```
Comment (bwplotka) [inline docs/components/compact.md:15]:
```suggestion
- creating 1h downsampling for blocks larger than **3 days** (2w for Thanos Compactions).
```

File: docs/components/compact.md [hunk 2]
```diff
@@ -12,7 +12,7 @@ It is generally not semantically concurrency safe and must be deployed as a sing
 It is also responsible for downsampling of data:
 
 - creating 5m downsampling for blocks larger than **40 hours** (2d, 2w)
```
Comment (bwplotka) [inline docs/components/compact.md:14]:
```suggestion
- creating 5m downsampling for blocks larger than **40 hours** (2d or 2w for Thanos compactions)
```

File: cmd/thanos/compact.go [hunk 3]
```diff
@@ -344,6 +344,13 @@ func runCompact(
 		level.Warn(logger).Log("msg", "Max compaction level is lower than should be", "current", maxCompactionLevel, "default", compactions.maxLevel())
 	}
 
+	// Ensure the lowest resolution downsampling is triggered at the highest compaction level.
+	downsample_range1 := downsample.DownsampleRange1 * time.Millisecond
+	if downsample_range1 <= compactions[len(compactions)-2] ||
+		downsample_range1 >= compactions[len(compactions)-1] {
+		panic("DownsampleRange1 must be between the final two compaction levels")
```
Comment (bwplotka) [inline cmd/thanos/compact.go:351]: We should never ever panic: https://thanos.io/contributing/coding-style-guide.md/#never-use-panics

What we are trying to achieve here? `compactions` are hardcoded so we can totally test this on unit test, no?

File: pkg/compact/downsample/downsample.go [hunk 4]
```diff
@@ -33,8 +33,11 @@ const (
 
 // Downsampling ranges i.e. minimum block size after which we start to downsample blocks (in seconds).
 const (
-	DownsampleRange0 = 40 * 60 * 60 * 1000      // 40 hours.
-	DownsampleRange1 = 10 * 24 * 60 * 60 * 1000 // 10 days.
+	// NOTE: although DownsampleRange1 can be as high as our last compaction level (14d),
+	//   we use just over the penultimate level to accommodate compacted blocks migrated
```
Comment (bwplotka) [inline pkg/compact/downsample/downsample.go:37]: `penultimate`? :thinking:

File: pkg/compact/downsample/downsample.go [hunk 5]
```diff
@@ -33,8 +33,11 @@ const (
...
+	//   from other systems (e.g. Prometheus server, which maxes out at 6.75d).
+	DownsampleRange0 = 40 * 60 * 60 * 1000     // 40 hours.
+	DownsampleRange1 = 3 * 24 * 60 * 60 * 1000 // 3 days.
```
Comment (bwplotka) [inline pkg/compact/downsample/downsample.go:40]: let's maybe use this occasion and simplify naming a bit and make it more clear.
```suggestion
	MinBlockRangeForResLevel1 = 40 * 60 * 60 * 1000     // 40 hours.
	MinBlockRangeForResLevel2 = 3 * 24 * 60 * 60 * 1000 // 3 days.
```

What to do  you think? :hugs:

File: cmd/thanos/compact.go (hunk 3 above)
Comment (belm0) [reply cmd/thanos/compact.go:351]: This code is asserting an invariant between two internal constants.  It's a developer-facing panic, users will never see it.

The program should not run if these constants are misconfigured.  Using a runtime error is a stronger safeguard against that than a unit test.

Thanos codebase uses panic elsewhere, e.g. https://github.com/thanos-io/thanos/blob/2634e0331487248becb0b1e7c49f88aa533e2bb7/cmd/thanos/main.go#L86

File: pkg/compact/downsample/downsample.go (hunk 5 above)
Comment (belm0) [reply pkg/compact/downsample/downsample.go]: done

File: pkg/compact/downsample/downsample.go (hunk 4 above)
Comment (belm0) [reply pkg/compact/downsample/downsample.go]: changed to "2nd-to-last"

File: docs/components/compact.md (hunk 2 above)
Comment (belm0) [reply docs/components/compact.md]: done

**Comment by belm0** [comment]
> Also let's look on UntilNextDownsampling function which is totally broken after the change here

UntilNextDownsampling was already broken, because it assumes that the downsample limits exactly match the compaction schedule.

Yes, the misreporting will be more severe with the new setting.  Is it only reported by 'bucket inspect`, and does anyone use this info?  In my experience I found it useless and misleading because it didn't consider that compactions may never happen.

Anyway, I've added a TODO in the code about the problem.  Really I think the system should report an *accurate* UntilNextCompaction-- that would help users more.  And it should have unit test coverage...

**Comment by bwplotka** [comment]
> Really I think the system should report an accurate UntilNextCompaction instead-- that would be more generally useful. And it should have unit test coverage...

How it would be useful?

**Comment by belm0** [comment]
> How it would be useful?

UntilNextCompaction could be surfaced on the bucket UI for each block and help users understand the compaction schedule.  It would consider `compaction.failed` meta flag, as well as dependencies on surrounding blocks (e.g. a block in a gap that may never be scheduled for compaction).

And once there is good visibility for next compaction, surfacing next downsample is trivial...

**Comment by bwplotka** [comment]
>> How it would be useful?

> UntilNextCompaction could be surfaced on the bucket UI for each block and help users understand the compaction schedule. It would consider compaction.failed meta flag, as well as dependencies on surrounding blocks (e.g. a block in a gap that may never be scheduled for compaction).

Nice, still it would be nice to know the answer on ... why. (: Why users need to understand the exact compaction schedule? We run huge Thanos systems (I am actually on-call from tomorrow) and I don't care, and don't want to care when compaction happens as long as it eventually happen. So yea, I cannot see the reasons (: 

But maybe I am missing something :hugs: Let us know.

> And once there is good visibility for next compaction, surfacing next downsample is trivial...

Well not really (: If there is constantly something to compact (because of 1000 sources pushing to same bucket), Downsampling will never happen.

**Comment by belm0** [comment]
> Nice, still it would be nice to know the answer on ... why. (: Why users need to understand the exact compaction schedule?

I'm pretty sure I wouldn't have spent days diagnosing #2413 if this had existed (and I'm no slouch at diagnosing problems).

**Comment by bwplotka** [comment]
Cool, but looks like the general documentation is missing on how it exactly
works, not a tooling problem no?

**Comment by bwplotka** [comment]
I mean, knowing exact compact time is not a root cause here (:

**Comment by belm0** [comment]
> Cool, but looks like the general documentation is missing on how it exactly works, not a tooling problem no?

If the bucket UI would show an indicator that "this block has no ETA on further compaction", then it provides a cue to the user and he can investigate why that is.

But I think we've deviated from the PR at hand at this point.  The TODO I added doesn't suggest "UntilNextCompaction", that's just my suggestion in this conversation.

I believe this PR is a beneficial, low risk, well-documented change-- ready to go.

**Pull request closed without merging**